acinclude/lib-checks.m4

```diff
@@ -259,6 +259,35 @@ return 0;
   SQUID_STATE_ROLLBACK(check_const_SSL_CTX_sess_set_get_cb)
 ])
 
+dnl Checks whether the X509_get0_signature() has const arguments
+AC_DEFUN([SQUID_CHECK_OPENSSL_CONST_X509_GET0_SIGNATURE_ARGS],[
+  AH_TEMPLATE(const_ASN1_BIT_STRING, "Define if X509_get0_signature() accepts const parameters")
+  AH_TEMPLATE(const_X509_ALGOR, "Define if X509_get0_signature() accepts const parameters")
+  SQUID_STATE_SAVE(check_const_X509_get0_signature_args)
+  AC_MSG_CHECKING("whether X509_get0_signature() accepts const parameters")
+  AC_COMPILE_IFELSE([AC_LANG_PROGRAM([
+#include <openssl/ssl.h>
+    ],[
+#if HAVE_LIBCRYPTO_X509_GET0_SIGNATURE
+        const ASN1_BIT_STRING *sig = nullptr;
+        const X509_ALGOR *sig_alg;
+        X509_get0_signature(&sig, &sig_alg, nullptr);
+#else
+#error Missing X509_get0_signature()
+#endif
+    ])
+  ],[
+   AC_DEFINE_UNQUOTED(const_ASN1_BIT_STRING, [const ASN1_BIT_STRING])
+   AC_DEFINE_UNQUOTED(const_X509_ALGOR, [const X509_ALGOR])
```

Comment:
I do not think we should lie about the actual constness of those types.
I suggest using something like `extracted_ASN1_BIT_STRING` and
`extracted_X509_ALGOR` names instead.

Reply:
I had similar doubts, but view this naming as a clear/obvious indication that we must treat it as const even if the const modifier is not explicitly present (the non-prefixed type name is always available for non-const uses). If we remove the const_ part we may forget its hidden const-ness and let code through the audit process that modifies the value. QA build checks dont help detecting such things when the old-library code paths are tested.

Reply:
The correct solution for this valid "we might modify the returned values" concern is to use real `const` arguments in the wrapper function profile. That way, the dangerous code will be confined to the wrapper function where the type name does not have to lie and where the audit is a lot more likely to identify problems (because of the tight and limited scope of the wrapper insides compared to various current and future callers).